scripts/src/java/org/oppia/android/scripts/kdoc/KDocCheck.kt

```diff
@@ -0,0 +1,284 @@
+  // checked for a KDoc.
+  val kDocNotRequiredAnnotationEntryList = listOf<String>(
+    "Rule",
+    "Mock",
+    "Test",
+    "Before",
+    "Captor",
+    "BeforeClass",
+    "After",
+    "AfterClass",
+    "Inject",
+    "Provides"
+  )
+
+  // A list of all the files to be exempted for this check.
+  val kdocExemptionList = loadKdocExemptionsProto(kdocExemptiontextProto).getExemptedFilePathList()
+
+  // A list of all kotlin files in the repo to be analyzed.
+  val searchFiles = RepositoryFile.collectSearchFiles(
+    repoPath = repoPath,
+    expectedExtension = ".kt",
+    exemptionsList = kdocExemptionList
+  )
+
+  // A list of all the files missing KDoc(s).
+  val matchedFiles = searchFiles.filter { file ->
+    hasKdocFailure(file, kDocNotRequiredAnnotationEntryList)
+  }
+
+  if (matchedFiles.isNotEmpty()) {
+    throw Exception("KDOC CHECK FAILED")
+  } else {
+    println("KDOC CHECK PASSED")
+  }
+}
+
+/** IntelliJ Project object, which is required to generate a [KtFile] from a Kotlin file. */
+private val project by lazy {
+  val config = CompilerConfiguration()
+  config.put(CLIConfigurationKeys.MESSAGE_COLLECTOR_KEY, MessageCollector.NONE)
+  KotlinCoreEnvironment.createForProduction(
+    Disposer.newDisposable(),
+    config,
+    EnvironmentConfigFiles.JVM_CONFIG_FILES
+  ).project
+}
+
+/**
+ * Genrates a [KtFile] from a Kotlin file.
+ *
+ * @param codeString the string equivalent of the Kotlin file
+ * @param fileName name of the file
+ * @return the generated [KtFile] for the given file
+ */
+private fun createKtFile(codeString: String, fileName: String): KtFile {
+  return PsiManager.getInstance(project)
+    .findFile(LightVirtualFile(fileName, KotlinFileType.INSTANCE, codeString)) as KtFile
+}
+
+/**
+ * Checks whether a file has a KDoc presence failure.
+ *
+ * @param file the file to be checked
+ * @param kDocNotRequiredAnnotationEntryList the list of annotation entries which when present
+ *     on an element, the element does not needs to be checked for a KDoc.
+ * @return whether the file has a KDoc presence failure
+ */
+private fun hasKdocFailure(file: File, kDocNotRequiredAnnotationEntryList: List<String>): Boolean {
+  val ktFile = createKtFile(file.asString(), file.name)
+  return ktFile.children.fold(initial = false) { isFailing, elem ->
+    val childFailingKdoc = if (elem is KtElement) elementIterator(
+      elem,
+      file,
+      kDocNotRequiredAnnotationEntryList
+    ) else false
+    isFailing || childFailingKdoc
+  }
+}
+
+/**
+ * Returns the string equivalent of a file.
+ *
+ * @return the string equivalent
+ */
+private fun File.asString(): String {
+  val fileContents = StringBuilder(this.length().toInt())
+  Scanner(this).use { scanner ->
+    while (scanner.hasNextLine()) {
+      fileContents.append(scanner.nextLine() + System.lineSeparator())
+    }
+    return fileContents.toString()
+  }
+}
```

Comment:
Could we use [``File.readText()``](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/java.io.-file/read-text.html) instead?

Reply:
Actually, if we use `File.readText()`, then it gives this error while parsing one of the files because of the line separator, hence I had to go by this approach.

```
Exception in thread "main" java.lang.AssertionError: Wrong line separators: '....gae.model\r\n\r\nimport...' at offset 49
        at org.jetbrains.kotlin.com.intellij.openapi.util.text.StringUtil.assertValidSeparators(StringUtil.java:2450)
        at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.assertValidSeparators(DocumentImpl.java:690)
        at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.<init>(DocumentImpl.java:119)
        at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.<init>(DocumentImpl.java:114)
        at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.<init>(DocumentImpl.java:105)
        at org.jetbrains.kotlin.com.intellij.core.CoreApplicationEnvironment.lambda$new$1(CoreApplicationEnvironment.java:94)
        at org.jetbrains.kotlin.com.intellij.mock.MockFileDocumentManagerImpl.getDocument(MockFileDocumentManagerImpl.java:42)
        at org.jetbrains.kotlin.com.intellij.psi.AbstractFileViewProvider.getDocument(AbstractFileViewProvider.java:180)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.retrieveLineNumberForElement(KDocValidityCheck.kt:260)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.checkIfKDocMissing(KDocValidityCheck.kt:208)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.elementIterator(KDocValidityCheck.kt:169)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.hasKdocFailure(KDocValidityCheck.kt:130)
        at org.oppia.android.scripts.docs.KDocValidityCheckKt.main(KDocValidityCheck.kt:86)
```

Reply:
Ah. You could maybe do something like:

```kotlin
file.readLines().joinToString(separateor = System.lineSeparator())
```

Would that work?